### PR TITLE
LPE switch to velocity based optical flow

### PR DESCRIPTION
**Pull request by jgoppert**
This switches LPE from treating optical flow as a position measurement and keeping the integrated state to just treating it as a velocity measurement.  This makes it work more easily with external vision measurements. I just flew this in position hold at 100 ft with only sonar, so it is very robust. I also manually increased the px4flow autoexposure maximum to get this kind of performance, for some reason the default is set very low.

I've also been flying this branch in gazebo simulations with external vision measurements and it works well.

@mhkabir probably will be useful to you

### CODE REVIEW

**Comment by mhkabir** [comment]
Thanks! 

@Aperture-Laboratories You will want this ;)

File: src/modules/local_position_estimator/sensors/vision.cpp [hunk 1]
```diff
@@ -6,8 +6,8 @@ extern orb_advert_t mavlink_log_pub;
 
 // required number of samples for sensor
 // to initialize
-static const uint32_t 		REQ_VISION_INIT_COUNT = 20;
-static const uint32_t 		VISION_TIMEOUT =    500000;	// 0.5 s
+static const uint32_t 		REQ_VISION_INIT_COUNT = 5;
+static const uint32_t 		VISION_TIMEOUT =    10000000;	// 10 s
```
Comment (ecmnet) [inline src/modules/local_position_estimator/sensors/vision.cpp]: @jgoppert Switching Flow to velocity also helps me a lot e.g. in the case of indoors/outdoors transition (thanks for that), when onboard vision might not be available anymore. I am just curious, why VISION_TIMEOUT now is set to 10s - indeed I reduced that to 0.1 seconds to let LPE know as soon as possible that vision is not available anymore.

File: src/modules/local_position_estimator/sensors/vision.cpp [hunk 2]
```diff
@@ -21,7 +21,6 @@ void BlockLocalPositionEstimator::visionInit()
 
 	// increament sums for mean
 	if (_visionStats.getCount() > REQ_VISION_INIT_COUNT) {
-		_visionOrigin = _visionStats.getMean();
```
Comment (ecmnet) [inline src/modules/local_position_estimator/sensors/vision.cpp]: @jgoppert What is the idea behind removing visionOrigin?

**Comment by jgoppert** [comment]
@ecmnet I assume that the vision system knows it's origin and is giving measements relative to that. The offset it only necessary if there are two independent position measurements and since optical fow is eliminated it removes the requirement for that.

**Comment by jgoppert** [comment]
Vision timeout is set to 10 to prevent having to reinitialize it, which should be unnecessary. The vxy pub will still trigger and drop you to altitude node if you don't have vision data for awhile.

**Comment by ecmnet** [comment]
@jgoppert What you mentioned are assumptions, which do not apply at least to my vision implementation. Fr example, in case of rapid rotation, it looses  track and needs to be re-initialized, which means a new origin. In case of those vision resets, I don't want to drop to altitude, but continue position control with re-initialized vision (which, btw. works twice as accurate as flow indoors).
Maybe I mixed up 'vision' with 'mocap' as I am not quite sure what the different definitions are.

**Comment by jgoppert** [comment]
@ecmnet, we will need to add a vision based velocity correction back for that case

**Comment by ecmnet** [comment]
@jgoppert I do not really understand how this would solve the issue. I case of vision failure, the best guess for local position when vision resumes with a new coordinate frame, is the LPE position.

**Comment by jgoppert** [comment]
Yes, if we make the vision measurement a velocity measurement it will have that behavior automatically.

**Comment by jgoppert** [comment]
@LorenzMeier @mhkabir @bkueng @ecmnet I think this is ready for merge, please review.

**Comment by ecmnet** [comment]
@jgoppert I saw, that this is flow related. Do you plan to adjust vision accordingly? If not you may consider my pull https://github.com/PX4/Firmware/pull/5482.

**Comment by jgoppert** [comment]
@ecmnet This code has actually been test flown with vision based position measurements. I think adding the velocity based position will have to be in a separate PR, it is important to me that we get it in though.

**Comment by ecmnet** [comment]
@jgoppert Are there any logs of those vision based flights available?

**Comment by jgoppert** [comment]
Yes, will upload some in a few hours. This log will be using vision with known landmarks so the data has no drift in it like you would get with visual odometry.

**Comment by ecmnet** [comment]
@jgoppert Ok. In that case you might not have the issue of re-initializing at an unknown position. For indoors flights, I dropped flow completely now relying on visual odometry only. With my patch, I have good results, even if visual timeouts occur.

**Comment by jgoppert** [comment]
I will add support for your setup on this pr tonight.

**Comment by jgoppert** [comment]
@ecmnet Try 06ff49f out. Just set LPE_VIS_MODE = 2 (VISION_VELOCITY)

**Comment by ecmnet** [comment]
@jgoppert I will give it a try - this changes a lot on my vision side. AM I correct to use now the MAVLINK message VISION_SPEED_ESTIMATE instead of VISION_POSITION_ESTIMATE?

**Comment by jgoppert** [comment]
No, would just send the same message but only set the velocity fields.

**Comment by jgoppert** [comment]
I didn't see a uorb topic for vision speed data, so didn't go that route.

**Comment by ecmnet** [comment]
@jgoppert VISION_POSITION_ESTIMATE I use has no velocity fields (might be that it is outdated):

 <message id="102" name="VISION_POSITION_ESTIMATE">
               <field type="uint64_t" name="usec">Timestamp (microseconds, synced to UNIX time or since system boot)</field>
               <field type="float" name="x">Global X position</field>
               <field type="float" name="y">Global Y position</field>
               <field type="float" name="z">Global Z position</field>
               <field type="float" name="roll">Roll angle in rad</field>
               <field type="float" name="pitch">Pitch angle in rad</field>
               <field type="float" name="yaw">Yaw angle in rad</field>
          </message>

**Comment by mhkabir** [comment]
Yes, for the speed message, you need to add the uorb plumbing.

**Comment by jgoppert** [comment]
@ecmnet @mhkabir   I got all of the plumbing done and created a new PR for this based on this PR here: https://github.com/PX4/Firmware/pull/5574

**Pull request merged**